src/node/internal/internal_timers.ts

```diff
@@ -0,0 +1,320 @@
+import { validateFunction, validateNumber } from 'node-internal:validators';
+import { ERR_OUT_OF_RANGE } from 'node-internal:internal_errors';
+
+// Timeout values > TIMEOUT_MAX are set to 1.
+const TIMEOUT_MAX = 2 ** 31 - 1;
+
+let clearTimeoutImpl: (obj: Timeout) => void;
+let clearImmediateImpl: (obj: Immediate) => void;
+
+function getTimerDuration(msecs: unknown, name: string): number {
+  validateNumber(msecs, name);
+  if (msecs < 0 || !Number.isFinite(msecs)) {
+    throw new ERR_OUT_OF_RANGE(name, 'a non-negative finite number', msecs);
+  }
+
+  // Ensure that msecs fits into signed int32
+  if (msecs > TIMEOUT_MAX) {
+    return TIMEOUT_MAX;
+  }
+
+  return msecs;
+}
+
+class Timeout {
+  #timer: string | number | this | undefined;
+  #callback: (...args: unknown[]) => unknown;
+  #after: number;
+  #args: unknown[];
+  #isRepeat: boolean;
+  #isRefed: boolean;
+
+  public constructor(
+    callback: (...args: unknown[]) => unknown,
+    after?: number,
+    args: unknown[] = [],
+    isRepeat: boolean = false,
+    isRefed: boolean = false
+  ) {
+    if (after === undefined) {
+      after = 1;
+    } else {
+      after *= 1; // Coalesce to number or NaN
+    }
+
+    this.#callback = callback;
+    this.#after = after;
+    this.#args = args;
+    this.#isRepeat = isRepeat;
+    this.#isRefed = isRefed;
+    this.#constructTimer();
+  }
+
+  #constructTimer(): void {
+    if (this.#isRepeat) {
+      // @ts-expect-error TS2322 Due to difference between Node.js and globals
+      this.#timer = globalThis.setInterval(
+        () => this.#callback(...this.#args),
+        this.#after
+      );
+    } else {
+      // @ts-expect-error TS2322 Due to difference between Node.js and globals
+      this.#timer = globalThis.setTimeout(
+        () => this.#callback(...this.#args),
+        this.#after
+      );
+    }
+  }
+
+  #clearTimeout(): void {
+    if (this.#isRepeat) {
+      globalThis.clearInterval(this.#timer);
+    }
+    {
+      globalThis.clearTimeout(this.#timer);
+    }
+  }
+
+  public refresh(): this {
+    this.#clearTimeout();
+    this.#constructTimer();
+    return this;
+  }
+
+  public unref(): this {
+    // TODO(soon): Implement this
+    this.#isRefed = false;
+    return this;
+  }
+
+  public ref(): this {
+    // TODO(soon): Implement this
+    this.#isRefed = true;
+    return this;
+  }
+
+  public hasRef(): boolean {
+    return this.#isRefed;
+  }
+
+  public close(): this {
+    this.#clearTimeout();
+    return this;
+  }
+
+  public [Symbol.dispose](): void {
+    this.#clearTimeout();
+  }
+
+  public [Symbol.toPrimitive](): number {
+    // @ts-expect-error TS2322 Timer is actually an ID.
+    return this.#timer;
+  }
+
+  static {
+    clearTimeoutImpl = (obj: Timeout): void => {
+      obj.#clearTimeout();
+    };
+  }
+}
+
+class Immediate {
```

Comment:
hmm... we already have the `Immediate` class implemented in c++ in support of the existing `setImmediate()` impl... is there really a need to implement it again here? That's likely to cause issues I think

Reply:
It doesn't seem to implement Symbol.dispose(). Do we implement SymbolDispose in anywhere in workerd in C++?

Reply:
Yep. Search for `JSG_DISPOSE`